File: frontend/src/types/ticket.ts
export type TicketCategory = 'IT' | 'Facilities' | 'HR'

export type TicketPriority = 'Low' | 'Medium' | 'High' | 'Urgent'

export type TicketStatus = 'Open' | 'In Progress' | 'Resolved' | 'Closed'

export interface Ticket {
  id: number
  title: string
  description: string
  category: TicketCategory
  priority: TicketPriority
  status: TicketStatus
  created_by: number
  created_at: string
  updated_at: string
}

export interface StatusHistoryEntry {
  id: number
  from_status: TicketStatus | null
  to_status: TicketStatus
  changed_by: number
  changed_at: string
}

export interface TicketDetail extends Ticket {
  history: StatusHistoryEntry[]
}

export interface PaginatedTickets {
  items: Ticket[]
  total: number
  page: number
  page_size: number
}

export interface TicketCreateInput {
  title: string
  description: string
  category: TicketCategory
  priority: TicketPriority
}

export interface TicketListParams {
  status?: TicketStatus
  category?: TicketCategory
  priority?: TicketPriority
  sort_by?: 'created_at' | 'priority'
  order?: 'asc' | 'desc'
  page?: number
  page_size?: number
}

export const TICKET_CATEGORIES: TicketCategory[] = ['IT', 'Facilities', 'HR']

export const TICKET_PRIORITIES: TicketPriority[] = ['Low', 'Medium', 'High', 'Urgent']

export const TICKET_STATUSES: TicketStatus[] = ['Open', 'In Progress', 'Resolved', 'Closed']

export function nextStatus(current: TicketStatus): TicketStatus | null {
  const index = TICKET_STATUSES.indexOf(current)
  return index >= 0 && index + 1 < TICKET_STATUSES.length ? TICKET_STATUSES[index + 1] : null
}

File: backend/app/schemas/ticket.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import TicketCategory, TicketPriority, TicketStatus


class TicketCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str = Field(min_length=1, max_length=10_000)
    category: TicketCategory
    priority: TicketPriority


class TicketRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str
    category: TicketCategory
    priority: TicketPriority
    status: TicketStatus
    created_by: int
    created_at: datetime
    updated_at: datetime


class StatusHistoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    from_status: TicketStatus | None
    to_status: TicketStatus
    changed_by: int
    changed_at: datetime
    resolution_note: str | None


class TicketDetail(TicketRead):
    history: list[StatusHistoryRead]


class TicketStatusUpdate(BaseModel):
    status: TicketStatus
    resolution_note: str | None = None


class PaginatedTickets(BaseModel):
    items: list[TicketRead]
    total: int
    page: int
    page_size: int
